cmd/nilaway/main.go

```diff
@@ -18,10 +18,105 @@
 package main
 
 import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
 	"go.uber.org/nilaway"
+	"go.uber.org/nilaway/config"
+	"golang.org/x/tools/go/analysis"
 	"golang.org/x/tools/go/analysis/singlechecker"
 )
 
+// Analyzer is identical to the one in nilaway.go, except that it overrides the run function for
+// extra filtering of errors, since the singlechecker does not support error suppression like other
+// popular linter drivers.
+var Analyzer = &analysis.Analyzer{
+	Name:       nilaway.Analyzer.Name,
+	Doc:        nilaway.Analyzer.Doc,
+	Run:        run,
+	FactTypes:  nilaway.Analyzer.FactTypes,
+	ResultType: nilaway.Analyzer.ResultType,
+	Requires:   nilaway.Analyzer.Requires,
+}
+
+var (
+	// _includeErrorsInFiles is a driver flag for specifying the list of file prefixes to only report errors.
+	_includeErrorsInFiles string
+	// _excludeErrorsInFiles is a driver flag for specifying the list of file prefixes to not report errors.
+	_excludeErrorsInFiles string
+)
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	// NilAway by default analyzes all packages, including dependencies. Even if specified to
+	// exclude packages from analysis via configurations, NilAway can still report errors on
+	// packages that are not analyzed if the nilness flow happens within the analyzed package, but
+	// the flow concerns a struct that is in an excluded package. The usual way to handle them is
+	// to suppress them at the driver level, but singlechecker does not support that yet. Therefore,
+	// here we add extra logic to filter the errors.
+
+	// Properly parse the error suppression flags.
+	var files [2][]string
+	for i, flagVal := range [...]string{_includeErrorsInFiles, _excludeErrorsInFiles} {
+		if flagVal == "" {
+			files[i] = nil
+			continue
+		}
+
+		// Convert the file paths to absolute paths.
+		list := strings.Split(flagVal, ",")
+		for i := range list {
+			p, err := filepath.Abs(list[i])
+			if err != nil {
+				return nil, fmt.Errorf("incorrect file path: %s", list[i])
+			}
+			list[i] = p
+		}
+		files[i] = list
+	}
+	includes, excludes := files[0], files[1]
+
+	// Override the report function to add error filtering logic.
+	report := pass.Report
+	pass.Report = func(d analysis.Diagnostic) {
+		p := pass.Fset.File(d.Pos).Name()
+		for _, e := range excludes {
+			if strings.HasPrefix(p, e) {
+				return
+			}
+		}
+
+		for _, i := range includes {
+			if strings.HasPrefix(p, i) {
+				report(d)
+				return
+			}
+		}
+	}
+
+	// Delegate the real analysis run to the original nilaway analyzer.
+	return nilaway.Analyzer.Run(pass)
+}
+
 func main() {
-	singlechecker.Main(nilaway.Analyzer)
+	// For better UX, we lift the flags from config.Analyzer to the top level so that users can
+	// specify them without having to specify the analyzer name ("nilaway_config").
```

Comment:
Sorry, not sure I understand what this is doing. Does this mean we can pass `nilaway:flagName` in the command line and it will be passed to `nilaway_config`, or just `--flagName`? Can you give an example here on how this looks from the user point of view?

Reply:
Without this, users will have to understand our internal architecture and pass the flags specifically to the "nilaway_config" analyzer:

```
nilaway -nilaway_config.flag1 <VALUE1> -nilaway_config.flag2 <VALUE2>
```

With the lifting logic here, this will become transparent to the users:

```
nilaway -flag1 <VALUE1> -flag2 <VALUE2>
```

and internall we pass those flags to the "nilaway_config" analyzer. 

I have updated the comments here to include this example (hopefully it makes it clearer)